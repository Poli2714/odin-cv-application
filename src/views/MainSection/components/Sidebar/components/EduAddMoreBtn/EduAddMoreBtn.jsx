import { useContext } from 'react';
import { v4 as uuidv4 } from 'uuid';

import { AddMoreBtn } from 'src/components/buttons';
import {
  EducationContext,
  EduErrMsgContext,
  KeyContext,
} from 'src/hooks/Contexts';
import { isEduFormValid } from 'src/utils';

function EduAddMoreBtn() {
  const { educations, setEducations } = useContext(EducationContext);
  const { setShowEduErrMessage } = useContext(EduErrMsgContext);
  const { activeKey, setActiveKey } = useContext(KeyContext);

  const handleClickAddEdu = () => {
    const lastKey = Object.keys(educations).at(-1);
    if (isEduFormValid(educations[lastKey])) {
      const newEduID = uuidv4();
      setEducations({
        ...educations,
        [newEduID]: {
          degree: '',
          university: '',
          eduStartDate: '',
          eduEndDate: '',
          eduInfo: '',
        },
      });
      setActiveKey({ ...activeKey, eduActiveKey: newEduID });
      setShowEduErrMessage(false);
    } else {
      activeKey.eduActiveKey !== lastKey &&
        setActiveKey({ ...activeKey, eduActiveKey: lastKey });
      setShowEduErrMessage(true);
    }
  };

  return <AddMoreBtn onClick={handleClickAddEdu} />;
}

export default EduAddMoreBtn;
