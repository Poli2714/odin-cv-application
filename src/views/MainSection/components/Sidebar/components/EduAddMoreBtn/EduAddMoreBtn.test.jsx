import { expect, test, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import { useState } from 'react';

import EduAddMoreBtn from './EduAddMoreBtn';
import {
  EducationContext,
  EduErrMsgContext,
  KeyContext,
} from 'src/hooks/Contexts';

vi.mock('src/components/buttons', () => ({
  AddMoreBtn: vi.fn(({ onClick }) => (
    <button onClick={onClick}>Mock Add</button>
  )),
}));

function MockParentComponent() {
  const [educations, setEducations] = useState({ fakeID: { someData: '' } });
  const [activeKey, setActiveKey] = useState({ eduActiveKey: '' });
  const [showEduErrMessage, setShowEduErrMessage] = useState(false);

  return (
    <EducationContext.Provider value={{ educations, setEducations }}>
      <KeyContext.Provider value={{ activeKey, setActiveKey }}>
        <EduErrMsgContext.Provider
          value={{ showEduErrMessage, setShowEduErrMessage }}
        >
          <EduAddMoreBtn />
        </EduErrMsgContext.Provider>
      </KeyContext.Provider>
    </EducationContext.Provider>
  );
}

test('renders EduAddMoreBtn', () => {
  render(<MockParentComponent />);
  expect(screen.getByRole('button')).toMatchSnapshot();
});
