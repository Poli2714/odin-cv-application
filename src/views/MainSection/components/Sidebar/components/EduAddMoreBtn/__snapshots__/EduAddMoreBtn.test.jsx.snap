// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders EduAddMoreBtn 1`] = `
<button>
  Mock Add
</button>
`;
